daemon/src/editor_connection.rs

```diff
@@ -0,0 +1,306 @@
+use std::{
+    collections::HashMap,
+    env,
+    path::{Path, PathBuf},
+};
+
+use anyhow::{bail, Context};
+use tracing::debug;
+
+use crate::{
+    ot::OTServer,
+    sandbox,
+    types::{
+        EditorProtocolMessageError, EditorProtocolMessageFromEditor, EditorProtocolMessageToEditor,
+        InsideMessage,
+    },
+};
+
+/// Represents a connection to an editor. Handles the OT. To keep the code testable and sync, we do
+/// the actual sending of messages in the daemon, and the functions here just *calculate* them.
+pub struct EditorConnection {
+    id: String,
+    // TODO: Feels a bit duplicated here?
+    base_dir: PathBuf,
+    /// There's one OTServer per open buffer.
+    ot_servers: HashMap<String, OTServer>,
+}
+
+impl EditorConnection {
+    pub fn new(id: String, base_dir: &Path) -> Self {
+        Self {
+            id,
+            base_dir: base_dir.to_owned(),
+            ot_servers: HashMap::new(),
+        }
+    }
+
+    pub fn owns(&self, file_path: &str) -> bool {
+        self.ot_servers.contains_key(file_path)
+    }
+
+    pub fn message_from_inside(
```

Comment:
This name is a bit confusing to me. This method seems to do two things:
- a side effect, but only for an edit
- generate a transformed message to the editor (but only for edits and cursor changes)

I'm wondering if it makes sense to split these two things, as it would be cool if the latter is side-effect free?

Reply:
The latter could then be done through an `InsideMessage -> Vec<EditorProtocolMessageToEditor>`-`From` trait. Ah no, it still needs the `rev_text_delta`, but that could be passed in.

Reply:
I think I was not aware, that the edit does the transformation as well.